### PR TITLE
fix: correct the display of the legend

### PR DESCRIPTION
**Pull request by Loori-R**
fix issue #93

### CODE REVIEW

**Review by dmitryk-dk — DISMISSED** [review]
LGTM!

**Review by dmitryk-dk — APPROVED** [review]
LGTM!

File: pkg/plugin/query.go [hunk 1]
```diff
@@ -119,13 +124,71 @@ func (q *Query) queryRangeURL(expr string, step time.Duration, queryParams url.V
...
+
+	customName := q.parseLegend(frame.Fields[1])
+	if customName != "" {
+		frame.Fields[1].Config = &data.FieldConfig{DisplayNameFromDS: customName}
+	}
+
+	frame.Name = customName
+}
+
+func metricNameFromLabels(f *data.Field) string {
```
Comment (hagen1778) [inline pkg/plugin/query.go:171]: This function is also confusing... The lack of description or tests makes me take guesses. And I get confused of `numLabels` usage...

File: pkg/plugin/query.go [hunk 2]
```diff
@@ -119,13 +124,71 @@ func (q *Query) queryRangeURL(expr string, step time.Duration, queryParams url.V
...
+			return ""
+		})
+		legend = result
+	}
+
+	// If legend is empty brackets, use query expression
 	if legend == "{}" {
 		return q.Expr
 	}
+
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply pkg/plugin/query.go]: logic was updated. made it simpler added tests

File: pkg/plugin/query.go (hunk 1 above)
Comment (dmitryk-dk) [reply pkg/plugin/query.go]: renamed function made it simpler, added tests

File: pkg/plugin/query_test.go [hunk 3]
```diff
@@ -180,3 +182,144 @@ func getTimeRage() TimeRange {
...
+			name:         "do not have label",
+			legendFormat: "{{job}}",
+			labels: data.Labels{
+				"instance": "127.0.0.1",
+			},
+			expr: "sum(vm_http_request_total)",
+			want: "sum(vm_http_request_total)",
+		},
+		{
+			name:         "has complex label",
```
Comment (hagen1778) [inline pkg/plugin/query_test.go:304]: what happens if we have `__name__` label only in the label set?

File: pkg/plugin/query.go [hunk 4]
```diff
@@ -119,13 +125,68 @@ func (q *Query) queryRangeURL(expr string, step time.Duration, queryParams url.V
...
+	frame.Name = customName
+}
+
+func labelsToString(labels data.Labels) string {
+	if labels == nil {
+		return "{}"
+	}
+
+	labelStrings := make([]string, 0, len(labels))
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply pkg/plugin/query.go]: updated

File: pkg/plugin/query.go [hunk 5]
```diff
@@ -119,13 +125,68 @@ func (q *Query) queryRangeURL(expr string, step time.Duration, queryParams url.V
...
+		}
+		labelStrings = append(labelStrings, fmt.Sprintf("%s=%q", label, value))
+	}
+
+	sort.Strings(labelStrings)
+	lbs := strings.Join(labelStrings, ",")
+	if lbs == "" {
+		return labels[metricsName]
+	}
+	return fmt.Sprintf("%s{%s}", labels[metricsName], lbs)
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply pkg/plugin/query.go]: updated

File: pkg/plugin/query.go [hunk 6]
```diff
@@ -119,13 +125,68 @@ func (q *Query) queryRangeURL(expr string, step time.Duration, queryParams url.V
...
+		if label == metricsName {
+			continue
+		}
+		labelStrings = append(labelStrings, fmt.Sprintf("%s=%q", label, value))
+	}
+
+	sort.Strings(labelStrings)
+	lbs := strings.Join(labelStrings, ",")
+	if lbs == "" {
+		return labels[metricsName]
```
[the remark this answers is not in the document]
Comment (dmitryk-dk) [reply pkg/plugin/query.go]: fixed, added check

File: pkg/plugin/query_test.go (hunk 3 above)
Comment (dmitryk-dk) [reply pkg/plugin/query_test.go:311]: added test

**Review by hagen1778 — APPROVED** [review]
LGTM

**Pull request merged**